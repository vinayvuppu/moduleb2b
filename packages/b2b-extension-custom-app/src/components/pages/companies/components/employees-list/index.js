export { default } from './employees-list';
export * from './employees-list';
