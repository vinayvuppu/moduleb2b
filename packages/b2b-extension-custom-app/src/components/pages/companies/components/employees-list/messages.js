import { defineMessages } from 'react-intl';

export default defineMessages({
  employeeNumber: {
    id: 'Company.Details.EmployeeTab.employeeNumber',
    defaultMessage: 'Employee Number',
  },
  email: {
    id: 'Company.Details.EmployeeTab.email',
    defaultMessage: 'Email',
  },
  firstName: {
    id: 'Company.Details.EmployeeTab.firstName',
    defaultMessage: 'First Name',
  },
  lastName: {
    id: 'Company.Details.EmployeeTab.lastName',
    defaultMessage: 'Last Name',
  },
  roles: {
    id: 'Company.Details.EmployeeTab.roles',
    defaultMessage: 'Roles',
  },
  budgetConsumed: {
    id: 'Company.Details.EmployeeTab.budgetConsumed',
    defaultMessage: 'Budget consumed',
  },
  noEmployeesTitle: {
    id: 'Company.Details.EmployeeTab.noEmployeesTitle',
    defaultMessage: 'There are no employees related to this company.',
  },
  fetchingEmployeesTitle: {
    id: 'Company.Details.EmployeeTab.fetchingEmployeesTitle',
    defaultMessage: 'Loading employees...',
  },
});
