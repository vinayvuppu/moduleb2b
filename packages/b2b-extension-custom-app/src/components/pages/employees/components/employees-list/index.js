export { default } from './employees-list';
