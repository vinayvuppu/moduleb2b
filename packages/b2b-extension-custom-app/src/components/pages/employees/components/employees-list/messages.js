import { defineMessages } from 'react-intl';

export default defineMessages({
  columnEmployeeNumber: {
    id: 'Employees.ListView.column.customerNumber',
    description: 'Title of the table column (customerNumber)',
    defaultMessage: 'Employee number',
  },
  columnExternalId: {
    id: 'Employees.ListView.column.externalId',
    description: 'Title of the table column (externalId)',
    defaultMessage: 'External ID',
  },
  columnEmail: {
    id: 'Employees.ListView.column.email',
    description: 'Title of the table column (email)',
    defaultMessage: 'Email',
  },
  columnFirstName: {
    id: 'Employees.ListView.column.firstName',
    description: 'Title of the table column (firstName)',
    defaultMessage: 'First name',
  },
  columnLastName: {
    id: 'Employees.ListView.column.lastName',
    description: 'Title of the table column (lastName)',
    defaultMessage: 'Last name',
  },
  columnCompanyName: {
    id: 'Employees.ListView.column.companyName',
    description: 'Title of the table column (companyName)',
    defaultMessage: 'Company',
  },
  columnCompany: {
    id: 'Employees.ListView.column.company',
    description: 'Title of the table column (company)',
    defaultMessage: 'Company',
  },
  columnCreatedAt: {
    id: 'Employees.ListView.column.createdAt',
    description: 'Title of the table column (createdAt)',
    defaultMessage: 'Created on',
  },
  columnLastModifiedAt: {
    id: 'Employees.ListView.column.lastModifiedAt',
    description: 'Title of the table column (lastModifiedAt)',
    defaultMessage: 'Modified on',
  },
  columnVatId: {
    id: 'Employees.ListView.column.vatId',
    description: 'Title of the table column (vatId)',
    defaultMessage: 'VAT ID',
  },
  columnStores: {
    id: 'Employees.ListView.column.stores',
    description: 'Title of the table column (stores)',
    defaultMessage: 'Store(s)',
  },
  storeKeyValueFallback: {
    id: 'Employees.ListView.column.value.storeKeyValueFallback',
    description: 'Fallback value for store column values',
    defaultMessage: '{key} (key)',
  },
  roles: {
    id: 'Employees.ListView.column.roles',
    defaultMessage: 'Roles',
  },
  budgetConsumed: {
    id: 'Employees.ListView.column.budgetConsumed',
    defaultMessage: 'Budget consumed',
  },
  noEmployeesTitle: {
    id: 'Employees.ListView.noEmployeesTitle',
    description: 'Title for no employees found.',
    defaultMessage: 'There are no employees in this project.',
  },
  noEmployeesDescriptionLink: {
    id: 'Employees.ListView.noEmployeesDescriptionLink',
    description: 'What to tell the user when no employees found.',
    defaultMessage: 'Add a employee.',
  },
  noResultsTitle: {
    id: 'Employees.ListView.noResultsTitle',
    description: 'Title for no (search) results found.',
    defaultMessage: 'No employees match these settings.',
  },
  noResultsDescription: {
    id: 'Employees.ListView.noResultsDescription',
    description: 'What to tell the user when no (search) results are found.',
    defaultMessage: 'Try changing the search settings.',
  },
  searchPlaceholder: {
    id: 'Employees.SearchByEmployees.searchPlaceholder',
    description: 'Placeholder label for the search input',
    defaultMessage: 'Search by email',
  },
  companyFilter: {
    id: 'Employees.ListView.filter.company',
    description: 'Label for the company filter option',
    defaultMessage: 'Employee company',
  },
  firstNameFilter: {
    id: 'Employees.ListView.filter.firstNameFilter',
    description: 'Label for the firstName filter option',
    defaultMessage: 'First name',
  },
  lastNameFilter: {
    id: 'Employees.ListView.filter.lastNameFilter',
    description: 'Label for the lastName filter option',
    defaultMessage: 'Last name',
  },
  vatIdFilter: {
    id: 'Employees.ListView.filter.vatIdFilter',
    description: 'Label for the vatId filter option',
    defaultMessage: 'VAT ID',
  },
  createdAtFilter: {
    id: 'Employees.ListView.filter.createdAtFilter',
    description: 'Label for the createdAt filter option',
    defaultMessage: 'Created',
  },
  lastModifiedAtFilter: {
    id: 'Employees.ListView.filter.lastModifiedAtFilter',
    description: 'Label for the lastModifiedAt filter option',
    defaultMessage: 'Modified',
  },
});
