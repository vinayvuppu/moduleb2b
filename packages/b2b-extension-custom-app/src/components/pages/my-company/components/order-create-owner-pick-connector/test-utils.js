/* eslint-disable import/prefer-default-export */
import { GRAPHQL_TARGETS } from '@commercetools-frontend/constants';
import flatMap from 'lodash.flatmap';
import intersection from 'lodash.intersection';
import FetchCustomersQuery from './order-create-customer-pick-connector.graphql';

export const createCustomersFetchQueryMock = (stores = []) => {
  const whereClause =
    stores.length === 0
      ? undefined
      : `(stores(key = "${stores.join(',')}") or stores is empty)`;
  const customers = [
    {
      id: 'customer-1',
      version: 4,
      firstName: 'Anna',
      lastName: 'Banana',
      email: 'annabanana@blablub.com',
      customerNumber: '111',
      companyName: null,
      customerGroup: null,
      stores: [
        {
          id: 'store-2',
          key: 'usa',
          nameAllLocales: [{ locale: 'en-GB', value: 'USA' }],
        },
      ],
      addresses: [
        {
          id: '_ZqTe0es',
          firstName: '3333',
          streetName: 'Fakestreet',
          streetNumber: '23',
          country: 'DE',
          additionalAddressInfo: 'Additional address info2',
          additionalStreetInfo: 'Additional street info',
          email: null,
          phone: null,
          city: null,
          company: null,
          lastName: null,
          postalCode: null,
          region: null,
        },
      ],
    },
    {
      id: 'customer-2',
      version: 1,
      firstName: 'Ali',
      lastName: 'Baba',
      customerNumber: '235',
      email: 'alibaba@testest.de',
      companyName: null,
      customerGroup: null,
      stores: [
        {
          id: 'store-1',
          key: 'germany',
          nameAllLocales: [{ locale: 'en-GB', value: 'Germany' }],
        },
      ],
      addresses: [],
    },
    {
      id: 'customer-3',
      version: 3,
      customerNumber: '128',
      firstName: 'Amin',
      lastName: 'Rocks',
      email: 'email@ct.de',
      companyName: null,
      customerGroup: null,
      stores: [],
      addresses: [
        {
          id: 'yUQGmqRG',
          firstName: 'first',
          lastName: 'last2',
          company: 'company',
          streetName: 'Street name',
          streetNumber: 'House number',
          city: 'city',
          postalCode: '66215',
          region: 'region',
          country: 'US',
          additionalAddressInfo: 'Additional address info2',
          additionalStreetInfo: 'Additional street info',
          email: 'email2@email.com',
          phone: 'phone',
        },
      ],
    },
  ].filter(customer => {
    // if we don't pass stores then we don't filter
    if (!stores) return true;

    const customerStoresArray = flatMap(customer.stores, s => s.key);
    return (
      // true if the customer belongs to one of the stores passed
      intersection(customerStoresArray, stores).length !== 0 ||
      // or the customer belongs to no store
      customerStoresArray.length === 0
    );
  });

  return {
    request: {
      variables: {
        target: GRAPHQL_TARGETS.COMMERCETOOLS_PLATFORM,
        projectKey: 'context-playground',
        limit: 20,
        locale: 'en',
        offset: 0,
        sort: ['createdAt desc'],
        where: whereClause,
      },
      query: FetchCustomersQuery,
    },
    result: {
      data: {
        customers: {
          count: 3,
          offset: 0,
          total: 3,
          results: customers,
        },
      },
    },
  };
};
