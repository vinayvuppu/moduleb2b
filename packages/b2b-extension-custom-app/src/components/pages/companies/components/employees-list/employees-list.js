import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, FormattedMessage } from 'react-intl';
import flowRight from 'lodash.flowright';

import { branchOnPermissions } from '@commercetools-frontend/permissions';
import {
  LoadingSpinner,
  Spacings,
  Table,
  Constraints,
} from '@commercetools-frontend/ui-kit';

import TabContentLayout from '@commercetools-local/core/components/tab-content-layout';
import Pagination from '@commercetools-local/core/components/search/pagination';
import PageBottomSpacer from '@commercetools-local/core/components/page-bottom-spacer';
import { formatMoney } from '@commercetools-local/utils/formats/money';

import { PERMISSIONS } from '../../../../../constants';
import EmployeesListConnector from '../employees-list-connector';
import B2BApolloClientContext from '../../../../common/b2b-apollo-client-context';
import { getRolByValue } from '../../../../utils/roles';

import messages from './messages';

const columnDefinitions = [
  {
    key: 'email',
    label: <FormattedMessage {...messages.email} />,
    flexGrow: 1,
  },
  {
    key: 'firstName',
    label: <FormattedMessage {...messages.firstName} />,
    flexGrow: 1,
  },
  {
    key: 'lastName',
    label: <FormattedMessage {...messages.lastName} />,
    flexGrow: 1,
  },
  {
    key: 'roles',
    label: <FormattedMessage {...messages.roles} />,
    flexGrow: 1,
  },
  {
    key: 'amountExpended',
    label: <FormattedMessage {...messages.budgetConsumed} />,
    flexGrow: 1,
  },
];

export class EmployeesList extends React.Component {
  static displayName = 'EmployeesList';

  static propTypes = {
    // regular props
    projectKey: PropTypes.string.isRequired,
    company: PropTypes.shape({
      id: PropTypes.string.isRequired,
      customerGroup: PropTypes.shape({
        id: PropTypes.string.isRequired,
      }),
    }),
    goToEmployeeDetails: PropTypes.func.isRequired,

    // HOC
    intl: PropTypes.shape({ formatMessage: PropTypes.func.isRequired }),
  };

  renderEmployeeRow = ({ rowIndex, columnKey }, employees) => {
    const employee = employees[rowIndex];
    let formattedValue = employee[columnKey];

    switch (columnKey) {
      case 'amountExpended':
        formattedValue = formatMoney(employee[columnKey], this.props.intl);
        break;
      case 'roles':
        formattedValue = (
          <div>
            {employee[columnKey].map(rol => (
              <div key={rol}>
                {getRolByValue(rol, this.props.intl.formatMessage).label}
              </div>
            ))}
          </div>
        );

        break;
      default:
        formattedValue = employee[columnKey];
        break;
    }

    return formattedValue;
  };

  handleOrderClick = employeeId => this.props.goToEmployeeDetails(employeeId);

  handleSortChange = (sortBy, sortDirection, searchQuery) => {
    searchQuery.set({
      ...searchQuery.value,
      page: 1,
      sorting: {
        key: sortBy,
        order: sortDirection.toLowerCase(),
      },
    });
  };

  render() {
    if (!this.props.company) {
      return <LoadingSpinner />;
    }
    return (
      <B2BApolloClientContext.Consumer>
        {({ apolloClient }) => {
          return (
            <EmployeesListConnector
              companyId={this.props.company.customerGroup.id}
              projectKey={this.props.projectKey}
              apolloClient={apolloClient}
            >
              {({ employeesFetcher, searchQuery }) => {
                if (employeesFetcher.isLoading)
                  return (
                    <Spacings.Stack scale="m" alignItems="center">
                      <LoadingSpinner>
                        <FormattedMessage
                          {...messages.fetchingEmployeesTitle}
                        />
                      </LoadingSpinner>
                    </Spacings.Stack>
                  );
                if (employeesFetcher.employees.results.length === 0)
                  return <FormattedMessage {...messages.noEmployeesTitle} />;

                return (
                  <TabContentLayout
                    data-testid="company-details-employee-tab"
                    data-track-component="CompanyEmployees"
                  >
                    <Constraints.Horizontal>
                      <Table
                        columns={columnDefinitions}
                        rowCount={employeesFetcher.employees.results.length}
                        items={employeesFetcher.employees.results}
                        itemRenderer={rowData =>
                          this.renderEmployeeRow(
                            rowData,
                            employeesFetcher.employees.results
                          )
                        }
                        shouldFillRemainingVerticalSpace={true}
                        onSortChange={(sortBy, sortDirection) => {
                          this.handleSortChange(
                            sortBy,
                            sortDirection,
                            searchQuery
                          );
                        }}
                        sortBy={searchQuery.value.sorting.key}
                        sortDirection={searchQuery.value.sorting.order.toUpperCase()}
                        onRowClick={(e, index) =>
                          this.handleOrderClick(
                            employeesFetcher.employees.results[index].id
                          )
                        }
                      >
                        <Pagination
                          totalItems={employeesFetcher.employees.total}
                          perPage={searchQuery.value.perPage}
                          page={searchQuery.value.page}
                          onPerPageChange={nextPerPage =>
                            searchQuery.set({
                              ...searchQuery.value,
                              perPage: nextPerPage,
                            })
                          }
                          onPageChange={nextPage =>
                            searchQuery.set({
                              ...searchQuery.value,
                              page: nextPage,
                            })
                          }
                        />
                      </Table>
                    </Constraints.Horizontal>
                    <PageBottomSpacer />
                  </TabContentLayout>
                );
              }}
            </EmployeesListConnector>
          );
        }}
      </B2BApolloClientContext.Consumer>
    );
  }
}

export default flowRight(
  branchOnPermissions([PERMISSIONS.ViewEmployees]),
  injectIntl
)(EmployeesList);
