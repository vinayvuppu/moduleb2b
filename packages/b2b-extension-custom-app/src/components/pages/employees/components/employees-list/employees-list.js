import PropTypes from 'prop-types';
import React from 'react';
import oneLineTrim from 'common-tags/lib/oneLineTrim';
import { getIn } from 'formik';
import flowRight from 'lodash.flowright';
import memoize from 'memoize-one';
import { injectIntl } from 'react-intl';
import { withRouter } from 'react-router-dom';
import {
  LoadingSpinner,
  Spacings,
  Table,
} from '@commercetools-frontend/ui-kit';
import { NO_VALUE_FALLBACK } from '@commercetools-frontend/constants';
import FormattedDateTime from '@commercetools-local/core/components/formatted-date-time';
import { formatCustomField } from '@commercetools-local/utils/formats';
import { injectTransformedLocalizedFields } from '@commercetools-local/utils/graphql';
import SearchViewControlledContainer from '@commercetools-local/core/components/search/search-view-controlled-container';
import ColumnManager from '@commercetools-local/core/components/column-manager';
import PageBottomSpacer from '@commercetools-local/core/components/page-bottom-spacer';
import { formatMoney } from '@commercetools-local/utils/formats/money';
import CustomFieldDefinitionsConnector from '@commercetools-local/core/components/custom-field-definitions-connector';
import { withApplicationContext } from '@commercetools-frontend/application-shell-connectors';
import { getRolByValue } from '../../../../utils/roles';
import EmployeesListCustomViewsConnector from '../employees-list-custom-views-connector';
import EmployeesViewLayout from '../employees-view-layout';
import EmployeesListConnector from '../employees-list-connector';
import { DEFAULT_PAGE_SIZES } from '../../../../constants/pagination';
import messages from './messages';
import { columnDefinitions as defaultColumnDefinitions } from './column-definitions';
import createFilterDefinitions from './filter-definitions';
import B2BApolloClientContext from '../../../../common/b2b-apollo-client-context';

const getColumns = memoize(visibleColumnsOfView => {
  const visibleColumns = visibleColumnsOfView
    ? visibleColumnsOfView
        .map(visibleColumn =>
          defaultColumnDefinitions.find(
            availableColumn => availableColumn.key === visibleColumn
          )
        )
        .filter(Boolean)
    : defaultColumnDefinitions;

  return {
    visible: visibleColumns,
    available: defaultColumnDefinitions,
  };
});

export class EmployeesList extends React.PureComponent {
  static displayName = 'EmployeesList';

  static propTypes = {
    projectKey: PropTypes.string.isRequired,
    location: PropTypes.object.isRequired,

    // withRouter
    history: PropTypes.shape({
      push: PropTypes.func.isRequired,
    }).isRequired,

    // withApplicationContext
    userId: PropTypes.string.isRequired,
    language: PropTypes.string.isRequired,
    languages: PropTypes.arrayOf(PropTypes.string).isRequired,

    // injectIntl
    intl: PropTypes.shape({ formatMessage: PropTypes.func.isRequired })
      .isRequired,
  };

  handleRowClick = (rowIndex, results) => {
    const employee = results[rowIndex];
    if (!employee) return;

    this.props.history.push(oneLineTrim`
      /${this.props.projectKey}
      /b2b-extension
      /employees
      /${employee.id}
    `);
  };

  handleUpdateColumns = columns =>
    this.setState({
      selectedColumns: columns.map(column => ({
        ...column,
        flexGrow: 1,
      })),
    });

  localizeStoreName = store =>
    injectTransformedLocalizedFields(store, [
      { from: 'nameAllLocales', to: 'name' },
    ]).name[this.props.language] ||
    this.props.intl.formatMessage(messages.storeKeyValueFallback, {
      key: store.key,
    });

  renderEmployeeRow = (
    results,
    { rowIndex, columnKey },
    customFieldDefinitions
  ) => {
    const value = results[rowIndex][columnKey];

    switch (columnKey) {
      case 'customerGroup':
        return value ? value.name : NO_VALUE_FALLBACK;
      case 'createdAt':
      case 'lastModifiedAt':
        return <FormattedDateTime type="datetime" value={value} />;
      case 'email':
      case 'firstName':
      case 'lastName':
      case 'externalId':
      case 'customerNumber':
        return value || NO_VALUE_FALLBACK;
      case 'stores':
        return (
          value.map(store => this.localizeStoreName(store)).join(', ') ||
          NO_VALUE_FALLBACK
        );
      case 'amountExpended':
        return formatMoney(value, this.props.intl);
      case 'roles':
        return (
          <div>
            {value.map(rol => (
              <div key={rol}>
                {getRolByValue(rol, this.props.intl.formatMessage).label}
              </div>
            ))}
          </div>
        );
      case 'companyName':
        return results[rowIndex].customerGroup?.name || NO_VALUE_FALLBACK;
      default: {
        // The key that identifies the custom field
        const customFieldDefinitionsOfColumnKey =
          customFieldDefinitions &&
          customFieldDefinitions.find(
            fieldDefinition => fieldDefinition.name === columnKey
          );

        // Because of using GraphQL, custom fields come now as `customFieldsRaw` which
        // is an object of name & value
        const customFieldsOfColumnKey = results[
          rowIndex
        ].custom?.customFieldsRaw?.find(
          customField =>
            customField.name === customFieldDefinitionsOfColumnKey?.name
        );

        return customFieldDefinitionsOfColumnKey && customFieldsOfColumnKey
          ? formatCustomField({
              value: customFieldsOfColumnKey.value,
              type: customFieldDefinitionsOfColumnKey.type,
              intl: this.props.intl,
              language: this.props.language,
              languages: this.props.languages,
            })
          : NO_VALUE_FALLBACK;
      }
    }
  };

  render() {
    const filterDefinitions = createFilterDefinitions({
      intl: this.props.intl,
    });

    return (
      <B2BApolloClientContext.Consumer>
        {({ apolloClient }) => {
          return (
            <EmployeesListCustomViewsConnector.Consumer>
              {({ activeView, setActiveView }) => (
                <EmployeesListConnector
                  projectKey={this.props.projectKey}
                  searchQuery={activeView}
                  apolloClient={apolloClient}
                >
                  {({ employeesFetcher }) => (
                    <CustomFieldDefinitionsConnector
                      resources={['customer']}
                      isDisabled={false}
                    >
                      {({ customFieldDefinitionsFetcher }) => (
                        <EmployeesViewLayout
                          projectKey={this.props.projectKey}
                          total={getIn(employeesFetcher.employees, 'total', 0)}
                        >
                          <SearchViewControlledContainer
                            key={activeView.id}
                            pageSizes={DEFAULT_PAGE_SIZES}
                            page={activeView.page}
                            perPage={activeView.perPage}
                            areFiltersVisible={true}
                            filterDefinitions={filterDefinitions}
                            count={employeesFetcher.employees.count}
                            total={employeesFetcher.employees.total}
                            results={employeesFetcher.employees.results}
                            onChange={nextFilters => {
                              setActiveView({
                                ...activeView,
                                ...nextFilters,
                              });
                            }}
                            value={activeView}
                            noResultsText={this.props.intl.formatMessage(
                              messages.noResultsTitle
                            )}
                            searchInputPlaceholder={this.props.intl.formatMessage(
                              messages.searchPlaceholder
                            )}
                          >
                            {({
                              rowCount,
                              results,
                              sorting,
                              onSortChange: setSorting,
                              measurementResetter,
                              footer,
                            }) => {
                              if (
                                employeesFetcher.isLoading ||
                                customFieldDefinitionsFetcher.isLoading
                              )
                                return (
                                  <Spacings.Stack scale="m" alignItems="center">
                                    <LoadingSpinner />
                                  </Spacings.Stack>
                                );
                              const columns = getColumns(
                                activeView.visibleColumns
                              );

                              return (
                                <Spacings.Stack scale="m">
                                  <ColumnManager
                                    availableColumns={columns.available}
                                    selectedColumns={columns.visible}
                                    onUpdateColumns={nextVisibleColumns => {
                                      const keysOfVisibleColumns = nextVisibleColumns.map(
                                        visibleColumn => visibleColumn.key
                                      );

                                      setActiveView({
                                        ...activeView,
                                        visibleColumns: keysOfVisibleColumns,
                                      });
                                    }}
                                  />
                                  <Table
                                    columns={columns.visible}
                                    itemRenderer={item =>
                                      this.renderEmployeeRow(
                                        results,
                                        item,
                                        customFieldDefinitionsFetcher.customFieldDefinitions
                                      )
                                    }
                                    rowCount={rowCount}
                                    onRowClick={(_, rowIndex) =>
                                      this.handleRowClick(rowIndex, results)
                                    }
                                    onSortChange={setSorting}
                                    sortBy={sorting.key}
                                    sortDirection={sorting.order.toUpperCase()}
                                    measurementResetter={measurementResetter}
                                    shouldFillRemainingVerticalSpace={true}
                                    items={results}
                                  >
                                    {footer}
                                    <PageBottomSpacer />
                                  </Table>
                                </Spacings.Stack>
                              );
                            }}
                          </SearchViewControlledContainer>
                        </EmployeesViewLayout>
                      )}
                    </CustomFieldDefinitionsConnector>
                  )}
                </EmployeesListConnector>
              )}
            </EmployeesListCustomViewsConnector.Consumer>
          );
        }}
      </B2BApolloClientContext.Consumer>
    );
  }
}

export default flowRight(
  injectIntl,
  withApplicationContext(applicationContext => ({
    language: applicationContext.dataLocale,
    languages: applicationContext.project.languages,
    userId: applicationContext.user.id,
  })),
  withRouter
)(EmployeesList);
