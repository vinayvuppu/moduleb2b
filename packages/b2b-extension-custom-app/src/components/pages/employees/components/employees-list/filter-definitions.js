import memoize from 'memoize-one';
import {
  createReferenceDefinitionsMap,
  createTextDefinitionsMap,
} from '@commercetools-local/core/components/search/standard-filter-definitions';
import messages from './messages';
import styles from './employees-list.mod.css';

export default memoize(({ intl }) => ({
  firstName: {
    label: intl.formatMessage(messages.firstNameFilter),
    filterTypes: createTextDefinitionsMap(intl),
  },
  lastName: {
    label: intl.formatMessage(messages.lastNameFilter),
    filterTypes: createTextDefinitionsMap(intl),
  },
  // vatId: {
  //   label: intl.formatMessage(messages.vatIdFilter),
  //   filterTypes: createTextDefinitionsMap(intl),
  // },
  // createdAt: {
  //   label: intl.formatMessage(messages.createdAtFilter),
  //   filterTypes: createDateDefinitionsMap(intl),
  // },
  // lastModifiedAt: {
  //   label: intl.formatMessage(messages.lastModifiedAtFilter),
  //   filterTypes: createDateDefinitionsMap(intl),
  // },
  'customerGroup.id': {
    label: intl.formatMessage(messages.companyFilter),
    filterTypes: createReferenceDefinitionsMap(intl, {
      type: 'customerGroup',
      isMulti: true,
      className: styles['customer-group-reference'],
    }),
  },
}));
