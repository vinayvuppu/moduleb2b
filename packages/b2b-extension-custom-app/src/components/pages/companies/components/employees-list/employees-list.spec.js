import PropTypes from 'prop-types';
import React from 'react';
import { shallow } from 'enzyme';
import { FormattedMessage } from 'react-intl';
import { LoadingSpinner } from '@commercetools-frontend/ui-kit';
import { intlMock } from '@commercetools-local/test-utils';
import EmployeesListConnector from '../employees-list-connector';
import { EmployeesList } from './employees-list';
import B2BApolloClientContext from '../../../../common/b2b-apollo-client-context';

jest.mock('@commercetools-local/utils/formats/date');

const createTestEmployee = props => ({
  id: 'abc-def',
  email: 'foo@bar.com',
  firstName: 'foo',
  lastName: 'bar',
  customerNumber: '1234',
  ...props,
});

const createTestProps = props => ({
  projectKey: 'test-project',
  company: {
    id: 'company-id-foo-123',
    customerGroup: { id: 'customer-group-id1' },
  },
  isActive: true,
  intl: intlMock,
  goToEmployeeDetails: jest.fn(),

  ...props,
});

const createConnectorProps = props => ({
  employeesFetcher: {
    isLoading: false,
    employees: {
      total: 1,
      count: 1,
      offset: 0,
      results: [createTestEmployee()],
    },
  },
  searchQuery: {
    get: jest.fn(),
    set: jest.fn(),
    value: {
      page: 1,
      perPage: 20,
      sorting: {
        key: 'createdAt',
        order: 'desc',
      },
    },
  },
  ...props,
});
const createB2BApolloClientContextProps = () => ({ apolloClient: {} });

const TestRowItem = ({ children }) => <div>{children}</div>;
TestRowItem.propTypes = {
  children: PropTypes.any.isRequired,
};

describe('rendering', () => {
  let props;
  let wrapper;
  describe('when loading', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<EmployeesList {...props} />);
      wrapper = wrapper
        .find(B2BApolloClientContext.Consumer)
        .renderProp('children')(createB2BApolloClientContextProps())
        .find(EmployeesListConnector)
        .renderProp('children')(
        createConnectorProps({
          employeesFetcher: {
            isLoading: true,
            employees: {
              total: 1,
              count: 1,
              offset: 0,
              results: [createTestEmployee()],
            },
          },
        })
      );
    });

    it('should render the loading spinner', () => {
      expect(wrapper).toRender(LoadingSpinner);
    });
  });

  describe('without employees', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<EmployeesList {...props} />);
      wrapper = wrapper
        .find(B2BApolloClientContext.Consumer)
        .renderProp('children')(createB2BApolloClientContextProps())
        .find(EmployeesListConnector)
        .renderProp('children')(
        createConnectorProps({
          employeesFetcher: {
            isLoading: true,
            employees: {
              total: 0,
              count: 0,
              offset: 0,
              results: [],
            },
          },
        })
      );
    });
    it('outputs correct tree', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render base elements', () => {
      expect(wrapper).toRender(FormattedMessage);
    });
  });

  describe('with employees', () => {
    const employees = [createTestEmployee()];
    let contentWrapper;
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<EmployeesList {...props} />);
      contentWrapper = wrapper
        .find(B2BApolloClientContext.Consumer)
        .renderProp('children')(createB2BApolloClientContextProps())
        .find(EmployeesListConnector)
        .renderProp('children')(createConnectorProps());
    });
    it('outputs correct tree', () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe('rendering the table', () => {
      it('should render table with employees', () => {
        expect(
          contentWrapper
            .find('Table')
            .prop('columns')
            .map(_ => _.key)
        ).toEqual([
          'email',
          'firstName',
          'lastName',
          'roles',
          'amountExpended',
        ]);
      });
      describe('columns rendering', () => {
        const index = 0;
        it('should render column email', () => {
          const key = 'email';
          const wrapperRow = shallow(
            <TestRowItem>
              {wrapper
                .instance()
                .renderEmployeeRow(
                  { rowIndex: index, columnKey: key },
                  employees
                )}
            </TestRowItem>
          );
          expect(wrapperRow).toHaveText('foo@bar.com');
        });
        it('should render column firstName', () => {
          const key = 'firstName';
          const wrapperRow = shallow(
            <TestRowItem>
              {wrapper
                .instance()
                .renderEmployeeRow(
                  { rowIndex: index, columnKey: key },
                  employees
                )}
            </TestRowItem>
          );
          expect(wrapperRow).toHaveText('foo');
        });
        it('should render column lastName', () => {
          const key = 'lastName';
          const wrapperRow = shallow(
            <TestRowItem>
              {wrapper
                .instance()
                .renderEmployeeRow(
                  { rowIndex: index, columnKey: key },
                  employees
                )}
            </TestRowItem>
          );
          expect(wrapperRow).toHaveText('bar');
        });
        it('should render column customerNumber', () => {
          const key = 'customerNumber';
          const wrapperRow = shallow(
            <TestRowItem>
              {wrapper
                .instance()
                .renderEmployeeRow(
                  { rowIndex: index, columnKey: key },
                  employees
                )}
            </TestRowItem>
          );
          expect(wrapperRow).toHaveText('1234');
        });
      });
    });
  });
  describe('callbacks', () => {
    describe('when a row is clicked', () => {
      beforeAll(() => {
        props = createTestProps();
        wrapper = shallow(<EmployeesList {...props} />);
        wrapper = wrapper
          .find(B2BApolloClientContext.Consumer)
          .renderProp('children')(createB2BApolloClientContextProps())
          .find(EmployeesListConnector)
          .renderProp('children')(createConnectorProps());
        wrapper.find('Table').prop('onRowClick')(null, 0);
      });

      it('should call `goToEmployeeDetails`', () => {
        expect(props.goToEmployeeDetails).toHaveBeenCalled();
      });
    });
  });
});
