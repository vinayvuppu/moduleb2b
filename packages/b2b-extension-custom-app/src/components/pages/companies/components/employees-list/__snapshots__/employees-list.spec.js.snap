// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering with employees outputs correct tree 1`] = `
<B2BApolloClientContextConsumer>
  <Component />
</B2BApolloClientContextConsumer>
`;

exports[`rendering without employees outputs correct tree 1`] = `
<Stack
  alignItems="center"
  scale="m"
>
  <LoadingSpinner
    scale="l"
  >
    <FormattedMessage
      defaultMessage="Loading employees..."
      id="Company.Details.EmployeeTab.fetchingEmployeesTitle"
      values={Object {}}
    />
  </LoadingSpinner>
</Stack>
`;
